rfd/0167-console-service.md

```diff
@@ -0,0 +1,225 @@
+---
+authors: Gabriel Corado (gabriel.oliveira@goteleport.com)
+state: draft
+---
+
+# RFD 0167 - Debug service
+
+## Required Approvals
+
+* Engineering: @r0mant && (@greedy52 || @rosstimothy)
+* Security: @reedloden || @jentfoo
+
+## What
+
+This RFD proposes a mechanism for Teleport administrators to switch log-level
+and enable profile consumption without restarting the instance.
+
+## Why
+
+During incidents in production environments, debug logs and diagnostic profiles
+(such as CPU and Memory allocation) are often unavailable for timely
+troubleshooting. To enable those, administrators need to restart the instance,
+potentially disrupting the application state and obscuring the issue they're
+trying to solve.
+
+In addition, the addition will make collecting profiles with a single command
+even on distroless environments where `curl` is not present.
+
+## Details
+
+A separate service will be introduced specifically for troubleshooting
+purposes. This will ensure the current diagnosis service is kept as is, not
+breaking current integrations or introducing behavior changes. The new service
+will also always be available (contrasting with the diagnosis service, which
+is optional and disabled by default), so in scenarios where users need to use
+it, they don't need to restart Teleport.
+
+In addition, the new service will listen into a Unix socket instead of TCP. This
+will make discoverability easier and discourage external usage, as the API is
+designed for internal usage.
+
+The `pprof` endpoints on diagnostics will still be available for existent
+integrations, such as the usage of
+`go tool pprof http://diag-addr/debug/pprof/profile`.
+
+### New service
+
+Teleport will start listening using a Unix socket located at
+`/var/run/teleport-diag-<node-id>.sock`.
+
+Having the Node ID on the socket name will also cover scenarios where multiple
+instances of running on the same machine exist. In this case, the consumers can
+rely on the Teleport configuration to locate the data directory and retrieve
+the ID.
+
+#### Endpoints
+
+In addition to the `pprof` endpoints, the service will also have endpoints that
```

Comment:
Have we considered requiring authentication and role specifications to be able to perform this debugging? The system level access is one protection, but it's offloading the security to our be our customers problem. It seems better if we can take on this boarding and control access through our RBAC as well.

Reply:
If I'm correct, that would require moving the endpoints to the auth service or the Web API, which I'm not sure is desired. Or at least duplicate the authentication/authorization logic in the service. In this case, we probably want to do the same thing as `tctl` commands, where it is possible to execute it from outside and inside (without credentials) the Teleport instances.

The original idea was to give administrators a command to quickly diagnose a problematic instance, making the command closer to the environment and always available (making it into the `teleport` CLI).